### ISSUE TITLE
Support multiple lfs.url configuration options

### ISSUE THREAD

**Issue by mathstuf** [git-lfs/git-lfs#3344]
It would be nice to support having multiple `lfs.url` settings in `.lfsconfig` for fetching. This would allow for a "backup" LFS server in case one has hit a rate limit to fall back to another instance.

Would uploads upload to all and only fail if all fail? Maybe it would only upload to the first URL.

**Comment by bk2204** [comment]
This is an interesting idea.  While I certainly see the value in it, I think my preference is for us to do the same thing Git does with multiple URLs if we add them (because that's less surprising behavior).  Git currently fetches from the first URL and pushes to all, and fails if pushing to any remote failed.

There is an existing issue (#3122) to handle rate limits more gracefully, which might solve the problem.  However, I don't have access to a server on which I can trigger rate limits, so while I'd like to fix it, I would only be able to simulate it in tests, which makes me a little nervous about trying a solution.

**Comment by mathstuf** [comment]
> I think my preference is for us to do the same thing Git does with multiple URLs if we add them (because that's less surprising behavior)

Hmm. I'll try and dig to see if I can find a rationale for that behavior in Git. My gut feeling for Git-LFS is that "I don't care where it comes from, I just want a copy of the data". But, that also applies to Git, so seeing the thought process would be interesting.

> However, I don't have access to a server on which I can trigger rate limits

GitHub has rate limits :) . They're not all that hard to hit either if you have any kind of CI enabled :/ .

**Comment by bk2204** [comment]
> Hmm. I'll try and dig to see if I can find a rationale for that behavior in Git. My gut feeling for Git-LFS is that "I don't care where it comes from, I just want a copy of the data". But, that also applies to Git, so seeing the thought process would be interesting.

For fetch URLs, I'm not sure why only the first one is honored.  I definitely can see the benefit to picking any option, but then again, the first option may be better in some way  (e.g. more up to date).  Then again, it could be just that we (upstream) have never considered any other option.

> GitHub has rate limits :) . They're not all that hard to hit either if you have any kind of CI enabled :/ .

Maybe I've been insufficiently skillful at trying to hit them.

**Comment by mathstuf** [comment]
> Maybe I've been insufficiently skillful at trying to hit them.

Well, it is used, so I guess it isn't "all that hard". The limit (without handing over cash) is 1GB of bandwidth / month. The main project I'm interested in this for has 250+ MB of assets and at least a dozen builders, nevermind dozens of developers all the data. The idea is that for LFS, one wants multiple sources of the data so that any one doesn't shut you out for going over some limit, but you can get it for your tests.

The end goal is for the build system to add rules to only download the data that is required for the tests that are actually enabled, so failing the build due to a rate limit seems harsh (especially if we can round-robin to another host).

**Comment by bk2204** [comment]
Okay, I think I may have misunderstood what you meant when you said “rate limits”.  I was thinking you were receiving a 429 from hitting the endpoint too fast, instead of hitting the data limits.  I apologize for sending you to #3122, which is definitely a separate issue.

Having thought about it some more, if we're just talking about fetches, I'm not opposed to trying multiple options in order, assuming that's achievable without too much work.  I think that might be appropriate to put behind an option in case upstream Git produces some different behavior for multiple fetch URLs.

Let me ping @git-lfs/core and see if anyone else has a strong view on the direction we should go on this.

**Comment by ttaylorr** [comment]
> Let me ping @git-lfs/core and see if anyone else has a strong view on the direction we should go on this.

I think it's a sensible idea, but we ought to mirror the ordering from Git in whatever our implementation might be in Git LFS. My understanding is the following:

1. When multiple `lfs.url`'s are present, and we are performing a 'fetch' operation, try _only_ from the first one.
2. When multiple `lfs.url`'s are present, and we are performing a 'push' operation, try all of them, and only exit cleanly iff every upstream takes the changes.

I have some concerns, but I'm not sure that any of them are showstoppers. That said, I think that we probably ought to discuss some of them a little further, so at the very least we understand what we're getting ourselves into.

For (1), I'm not sure that this is the most sensible behavior overall. I would think that we'd try (in some decidable ordering) to get all objects, and then whichever objects were missing would carry on to the next upstream, and so on until there were either no objects left (success), or no upstreams left (failure). I can't recall what Git does in this case, but if it's not this, I think it _may_ be worth deviating from.

For (2), I think that this makes sense, and is probably the best that we can do, but might be unsafe in general. Since we don't transact pushes (e.g., we can't "roll one back" if a later push fails), this can lead to uneven states between repositories. That might not matter, since we're only talking about sending objects somewhere, but it might matter, too. I'm not sure -- @bk2204 do you have any thoughts about this?

In general, though, I'm not sure that we need to limit this behavior to `lfs.url`'s only. Maybe we would do the same if there were multiple Git remotes (applying the usual inference rules to go from Git remote URL -> LFS server, or taking the `lfs.<remote>.url` configuration instead).

**Comment by bk2204** [comment]
For (1), Git ignores any remotes other than the first, which is, I believe, what this issue is suggesting we change.

So for (2), the failure case resulting in incomplete data is okay for Git, since it exits nonzero and prints an error.  I'm very familiar with that behavior in Git, and I think we should mirror it.  I don't recall at the moment if the first push fails, whether Git continues to push to the remainder, but we should mirror that as well.

**Comment by fstefanov** [comment]
git-lfs/2.4.0 (GitHub; linux amd64; go 1.9.4) always fetches from the second URL
```
cat .lfsconfig
[lfs]
url = "https//first-url.com/lfs"
url = "https//second-url.com/lfs"
```

**Comment by bk2204** [comment]
Right, because this is currently a single-valued option and the second value overrides the first one.

**Comment by sietse** [comment]
Another benefit to supporting multiple `lfs.url` options: it would allow merging two repositories – that started out as separate projects – into one repository. At present, if you `git fetch` the history of LFS-backed repository A into an existing LFS-backed repository B, you can't checkout commits from A's history because LFS gives "[404] Object [from A's history] does not exist on the [B's] server" errors.

Being able to specify a second LFS url would solve this, I think? Then LFS could try the second LFS URL if an object does not exist at the first LFS URL.

Steps to reproduce this use case:

- create two LFS-backed repositories, let's call them Aremote and Bremote.
- go to your local clone of Aremote.
- `git remote add Bremote https://git.example.com/Bremote`
- `git fetch Bremote`
- `git checkout Bremote/master`

That results in (linebreaks added):

> Error downloading object: myfile.pdf (20561ec):
> Smudge error:
> Error downloading myfile.pdf
> (20561ec37a8a7518480bf39d44a76ba86fa7800c382046310708abef993ac3f4):
> [20561ec37a8a7518480bf39d44a76ba86fa7800c382046310708abef993ac3f4]
> Object does not exist on the server or you don't have permissions to access it:
> [404] Object does not exist on the server or you don't have permissions to access it

**Comment by nhatkhai** [comment]
Could be possible that we have multiple one so we can either search for the object one by one on the list, or parallels when fetch?
My company having a mix-up repos connected together... each of then may have difference urls and this is difficult to having logic right.

**Comment by jhadida** [comment]
Would it be useful to use the same logic as git remotes with multiple push URLs possible, but only one fetch? 
```
git remote add <RemoteName> <FetchURL>
git remote set-url --add --push  <PushURL1>
git remote set-url --add --push  <PushURL2>
git remote set-url --add --push  <PushURL3>
...
```
That way no `.lfsconfig` would be required, we can push to multiple remotes with a single `git lfs push <RemoteName>` command, and `git lfs pull <RemoteName>` would use the single fetch URL. This seems explicit to me and I don't think it would lead to an ambiguous situation with either push/pull. And it also means I don't have to manage LFS URLs separately which I would consider a plus.

**Comment by bk2204** [comment]
There's some discussion of multiple push remotes in #1823, which outlines the work required to make that general case work.  It's necessary that some of the prep work there be done first to make it possible for people to use multiple remotes, since the current way the code is structured doesn't allow for that.

As for using multiple `lfs.url` entries that don't override each other, I'm pretty sure that would be a breaking change, so it would need to go into 4.0.  Otherwise, situations where a URL overrides another one would suddenly change to one where both are used, which isn't compatible.

**Comment by mathstuf** [comment]
Note that my preference is to push-to-at-least-one and fetch-from-any (because I don't care which remote provides it, just that it is provided). Otherwise the first entry will be unduly expected to have all objects rather than allowing fallback to others (e.g., say your GitHub remote has hit its bandwidth quota).
